### PR TITLE
Avoid empty hashes

### PR DESCRIPTION
**Pull request by Bugagazavr**
**Problem**: 

currently returned empty hash if model parser doesn't registered, swagger-ui doesn't check json object, for empty, and get a null, swagger-ui with swagger 2.0 spec doesn't support null's and not loaded.

**Solution**:

raise exception if parser for current model not registered or model parsing result contains empty hash, to specify this problem for developer

### CODE REVIEW

File: lib/grape-swagger/endpoint.rb [hunk 1]
```diff
@@ -232,17 +232,26 @@ def expose_params_from_model(model)
...
+        parser = klass.new(model, self)
 
         break
       end
 
-      @definitions[model_name] = { type: 'object', properties: properties || {} }
+      properties = parser.call unless parser.nil?
+
+      if parser.nil?
+        properties = { unsupported_model_parser: { type: :string, description: "No parser registed for #{model_name}" } }
```
Comment (dblock) [inline lib/grape-swagger/endpoint.rb]: If this is capitalized, it should be a full sentence, so ends with a period, please.

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,7 +8,8 @@
 #### Fixes
 
 * [#416](https://github.com/ruby-grape/grape-swagger/pull/416): Support recursive models - [@lest](https://github.com/lest).
-* [#418](https://github.com/ruby-grape/grape-swagger/pull/418): Replaced github ref to rubygems for external gems - [@Bugagazavr](https://github.com/Bugagazavr).
+* [#419](https://github.com/ruby-grape/grape-swagger/pull/419): Replaced github ref to rubygems for external gems - [@Bugagazavr](https://github.com/Bugagazavr).
+* [#420](https://github.com/ruby-grape/grape-swagger/pull/420): Avoid empty hashes, becuase this is a cause of some errors in swagger-ui - [@Bugagazavr](https://github.com/Bugagazavr).
```
Comment (dblock) [inline CHANGELOG.md]: I think this should be more specific, spell out the error that it avoids? 

Also, if this wasn't a problem in a previous release I would just remove this line.

**Comment by dblock** [comment]
Seems a little weird to get an error inside the model's JSON output. Shouldn't it just fail and error out if developer action is required?

**Comment by Bugagazavr** [comment]
> Seems a little weird to get an error inside the model's JSON output. Shouldn't it just fail and error out if developer action is required?

ok, im just add `raise GrapeSwagger::Errors::SwaggerSpec`, if something going wrong, ok?

**Comment by Bugagazavr** [comment]
@dblock done, added two exceptions `UnregistredParser` and `SwaggerSpec`, this should helps developers to specify a problem if something going wrong. 

Also added mock parser to passing tests without external parser library.

Thanks again, and excuse me for the lack of points in the end of sentences.

File: CHANGELOG.md [hunk 3]
```diff
@@ -8,7 +8,8 @@
 #### Fixes
 
 * [#416](https://github.com/ruby-grape/grape-swagger/pull/416): Support recursive models - [@lest](https://github.com/lest).
-* [#418](https://github.com/ruby-grape/grape-swagger/pull/418): Replaced github ref to rubygems for external gems - [@Bugagazavr](https://github.com/Bugagazavr).
+* [#419](https://github.com/ruby-grape/grape-swagger/pull/419): Replaced github ref to rubygems for external gems - [@Bugagazavr](https://github.com/Bugagazavr).
+* [#420](https://github.com/ruby-grape/grape-swagger/pull/420): Raise SwaggerSpec exception if swagger spec doesn't satified, when no parser for model registred or response model is empty - [@Bugagazavr](https://github.com/Bugagazavr).
```
Comment (dblock) [inline CHANGELOG.md]: Spelling: satiSfied, say "isn't satisfied".

File: lib/grape-swagger/endpoint.rb [hunk 4]
```diff
@@ -232,17 +232,23 @@ def expose_params_from_model(model)
...
-        properties = model_class.call
+        parser = klass.new(model, self)
 
         break
       end
 
-      @definitions[model_name] = { type: 'object', properties: properties || {} }
+      properties = parser.call unless parser.nil?
+
+      raise GrapeSwagger::Errors::UnregisteredParser, "No parser registred for #{model_name}." if parser.nil?
```
Comment (dblock) [inline lib/grape-swagger/endpoint.rb]: `unless parser` maybe?

File: lib/grape-swagger/endpoint.rb [hunk 5]
```diff
@@ -232,17 +232,23 @@ def expose_params_from_model(model)
...
+      raise GrapeSwagger::Errors::SwaggerSpec, "Empty model #{model_name}, swagger 2.0 doesn't support empty definitions." if properties.nil? || properties.empty?
```
Comment (dblock) [inline lib/grape-swagger/endpoint.rb]: `unless properties && properties.any?`

**Comment by dblock** [comment]
Very good. See comments. English is also my third language so no stress, you're doing great ;)

**Comment by Bugagazavr** [comment]
@dblock i'l merge this, ok?

**Comment by dblock** [comment]
Merged, looks good.

**Comment by stefan-kolb** [comment]
The extraction of the parser and introducing the GrapeSwagger::Errors::UnregisteredParser breaks usage for every user. Don't you think this should have been at least a minor version increase not only a patch from 0.20.3 to 0.20.4? To be honest, currently everything seems to break for me with every minor version. You are moving really really fast here.

**Comment by stefan-kolb** [comment]
Inside the https://github.com/ruby-grape/grape-swagger/blob/master/UPGRADING.md it is also refered to as a chang for 0.21.0

**Comment by stefan-kolb** [comment]
By the way to get me right, I don't think this is a good change for 0.20.3 -> 0.20.4. But it should instead be 0.21.0 as it breaks usage for customers!

**Comment by Bugagazavr** [comment]
Yes this is breaking changes.

@dblock should we yank this release and publish this changes as 0.21.0?

**Comment by dblock** [comment]
If you think it breaks everyone, yank it.

**Comment by LeFnord** [comment]
@Bugagazavr yeap, think it would be good to yank the release, not only of breaking changes, also the support for grape > 0.16.x was announced for 0.21.0 release, but before I want to fix most bugs, to have a (relatively) stable version, which works for most cases 😉 

think, we should have a better coordination 😄

**Comment by Bugagazavr** [comment]
Agreed, sorry for this, I yank 0.20.4 soon as possible

**Comment by LeFnord** [comment]
no problem,


but I think we should really have a better communication for coordination,

and maybe defining a roadmap, or similar ...


what did you think?


best


my private mail is: pscholz.le@gmail.com


Peter Scholz

Developer, SageOne

Sage


Office +49 (0)341 48440 3557

Twitter<https://twitter.com/sagegroupplc> | YouTube<https://www.youtube.com/user/SageGroupPlc> | LinkedIn<https://www.linkedin.com/company/sage-software>

________________________________
Von: Bugagazavr <notifications@github.com>
Gesendet: Dienstag, 17. Mai 2016 19:50:48
An: ruby-grape/grape-swagger
Cc: Scholz, Peter; Comment
Betreff: Re: [ruby-grape/grape-swagger] Avoid empty hashes (#420)

Agreed, sorry for this, I yank 0.20.4 soon as possible

Sent from my iPhone

> On 17 May 2016, at 20:44, peter scholz <notifications@github.com> wrote:
>
> @Bugagazavr yeap, think it would be good to yank the release, not only of breaking changes, also the support for grape > 0.16.x was announced for 0.21.0 release, but before I want to fix most bugs, to have a (relatively) stable version, which works for most cases ?
>
> think, we should have a better coordination ?
>
> -
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

-
You are receiving this because you commented.
Reply to this email directly or view it on GitHub<https://github.com/ruby-grape/grape-swagger/pull/420#issuecomment-219798904>

Sage Software GmbH, Emil-von-Behring-Strasse 8-14, 60439 Frankfurt am Main Sitz: Frankfurt am Main, Handelsregister Frankfurt am Main HRB 55497, Geschaeftsfuehrer: Michael Robinson, Dr. Thomas Scholtis

-----------------Disclaimer-----------------

Die in dieser E-Mail und den dazu gehoerigen Anhaengen (die Nachricht) enthaltenen Informationen sind nur fuer den Adressaten bestimmt und koennen vertrauliche und/oder rechtlich geschuetzte Informationen enthalten. Sollten Sie die Nachricht irrtuemlich erhalten haben, loeschen Sie die Nachricht bitte und benachrichtigen Sie den Absender, ohne die Nachricht zu kopieren oder zu verteilen oder ihren Inhalt an andere Personen weiterzugeben. Ausser bei Vorsatz oder grober Fahrlaessigkeit schliessen wir jegliche Haftung fuer Verluste oder Schaeden aus, die durch virenbefallene Software oder E-Mails verursacht werden.

-----------------Disclaimer-----------------

The information contained in this e-mail and any attachments (the message) is intended for the addressee only and may contain confidential and/or privileged information. If you have received the message by mistake please delete it and notify the sender and do not copy or distribute it or disclose its contents to anyone. Except in case of gross negligence or wilful misconduct we accept no liability for any loss or damage caused by software or e-mail viruses.

**Comment by dblock** [comment]
@LeFnord @Bugagazavr Feel free to improve the organizational aspects of this project in whichever way you see fit. However remember that the only thing we ask from contributors is their best effort, aka we ask nothing. Roadmaps tend to become commitments - I suggest replacing them with communicating intent via issues, READMEs and such instead.

**Comment by LeFnord** [comment]
@dblock this is meant with … the main goal should be that we go into the same direction 😉

**Pull request merged**